types/relay.go

```diff
@@ -36,3 +64,42 @@ type Relay struct {
 	CreatedAt                time.Time     `json:"createdAt"`
 	UpdatedAt                time.Time     `json:"updatedAt"`
 }
+
+func (r Relay) Validate() (err error) {
+	structType := reflect.TypeOf(r)
+	structVal := reflect.ValueOf(r)
+	fieldNum := structVal.NumField()
+
+	var isError bool
+
+	for i := 0; i < fieldNum; i++ {
+		field := structVal.Field(i)
+		fieldName := structType.Field(i).Name
+
+		isSet := field.IsValid() && !field.IsZero()
+
+		if isSet {
+			if fieldName == "IsError" {
+				isError = true
+			}
+
+			if shouldBeEmptyField[fieldName] || (!isError && errorField[fieldName]) {
+				return fmt.Errorf("%s should not be set", fieldName)
+			}
+
+			if fieldName == "ErrorType" && !validErrorTypes[field.String()] {
+				return fmt.Errorf("%s is not valid", fieldName)
+			}
+		}
+
+		if !isSet {
+			if shouldBeEmptyField[fieldName] || field.Kind() == reflect.Bool || (!isError && errorField[fieldName]) {
```

Comment:
A comment here on what is checking will be helpful for the future to minimize cognitive load